pkg/skaffold/schema/profiles.go

```diff
@@ -80,45 +90,78 @@ func checkKubeContextConsistency(contextSpecificProfiles []string, cliContext, e
 
 // activatedProfiles returns the activated profiles and activated profiles which are kube-context specific.
 // The latter matters for error reporting when the effective kube-context changes.
-func activatedProfiles(profiles []latest.Profile, opts cfg.SkaffoldOptions) ([]string, []string, error) {
-	var activated []string
-	var contextSpecificProfiles []string
+func activatedProfiles(profiles reflect.Value, version semver.Version, opts cfg.SkaffoldOptions) ([]string, []string, error) {
+	activated, contextSpecificProfiles, err := checkActivations(profiles, version, opts)
+	if err != nil {
+		return nil, nil, err
+	}
 
-	if opts.ProfileAutoActivation {
-		// Auto-activated profiles
-		for _, profile := range profiles {
-			for _, cond := range profile.Activation {
-				command := isCommand(cond.Command, opts)
+	for _, profile := range opts.Profiles {
+		if strings.HasPrefix(profile, "-") {
+			activated = removeValue(activated, strings.TrimPrefix(profile, "-"))
+		} else {
+			activated = append(activated, profile)
+		}
+	}
 
-				env, err := isEnv(cond.Env)
-				if err != nil {
-					return nil, nil, err
-				}
+	return activated, contextSpecificProfiles, nil
+}
+
+// checkActivations converts each profile `Activation` to a known version and applies the corresponding `checkActivation` function.
+func checkActivations(profiles reflect.Value, version semver.Version, opts cfg.SkaffoldOptions) ([]string, []string, error) {
+	if !opts.ProfileAutoActivation {
+		return nil, nil, nil
+	}
 
-				kubeContext, err := isKubeContext(cond.KubeContext, opts)
+	var activatedProfiles []string
+	var contextSpecificProfiles []string
+	v1b4, _ := apiversion.Parse(v1beta4.Version)
+	for i := 0; i < profiles.Len(); i++ {
+		activations := profiles.Index(i).FieldByName("Activation")
+		profileName := profiles.Index(i).FieldByName("Name").String()
+		if activations.IsValid() {
+			for j := 0; j < activations.Len(); j++ {
+				activation := activations.Index(j).Addr().Interface()
+				var isActive, isContextActivated bool
+				var err error
+				switch {
+				// Custom activation not supported before v1beta4
+				case version.LT(v1b4):
+					isActive, isContextActivated = true, false
+				// when modifying the `Activation` struct add a case condition here corresponding to the new version and a corresponding `checkActivation` function.
+				default:
+					var versionedActivation *v1beta4.Activation
```

Comment:
Aren't we better off cloning this to `latest.Activation`?  What if v2beta12 adds a new field to `Activation`, or a different activation type?

(I experimented with changing this and other uses to `latest.Activation` and it seemed to work fine for me.)

Reply:
So the idea here is that if `v2beta12` changes `Activation` struct then this code would change to:

```go
switch {
	// Custom activation not supported before v1beta4
	case version.LT(v1b4):
		isActive, isContextActivated = true, false
	// when modifying the `Activation` struct add a case condition here corresponding to the new version and a corresponding `checkActivation` function.
	case version.LT(v1b12):
		var versionedActivation *v1beta4.Activation
		skutil.CloneThroughJSON(activation, &versionedActivation)
		isActive, isContextActivated, err = checkActivationv1b4(versionedActivation, opts)
	default:
		var versionedActivation *v2beta12.Activation
		skutil.CloneThroughJSON(activation, &versionedActivation)
		isActive, isContextActivated, err = checkActivationb2v12(versionedActivation, opts)
	}
```

Reply:
If I changed this to `latest.Activation` like below:
```go
default:
	var versionedActivation *latest.Activation
	skutil.CloneThroughJSON(activation, &versionedActivation)
	isActive, isContextActivated, err = checkActivation(versionedActivation, opts)

func checkActivation(a *latest.Activation, opts cfg.SkaffoldOpts)
```

Then, the next person modifying `Activation` struct might erroneously just modify the `checkActivation` function without handling for the old versions. Doing it the current way also catches this error.

Reply:
Just so I understand, your way will error if there are new fields or renamed fields, then `CloneThroughJSON` should panic.

Reply:
discussed offline with @briandealwis. This approach requires devs to be aware of the context for this issue, which makes it susceptible to break in the future. Will rethink approach.